builder/openstack_id3/access_config.go

```diff
@@ -0,0 +1,150 @@
+package openstack_id3
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/mitchellh/packer/common"
+	"github.com/mitchellh/packer/packer"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
+)
+
+// AccessConfig is for common configuration related to openstack access
+type AccessConfig struct {
+	Username  string `mapstructure:"username"`
+	UserId    string `mapstructure:"user_id"`
+	Password  string `mapstructure:"password"`
+	ApiKey    string `mapstructure:"api_key"`
+	Project   string `mapstructure:"project"`
+	ProjectId string `mapstructure:"project_id"`
+	Provider  string `mapstructure:"provider"`
+	RawRegion string `mapstructure:"region"`
+	ProxyUrl  string `mapstructure:"proxy_url"`
+	TenantId  string `mapstructure:"tenant_id"`
+	Insecure  bool   `mapstructure:"insecure"`
+	Domain    string `mapstructure:"domain"`
+	DomainId  string `mapstructure:"domain_id"`
+}
+
+// Auth returns a ProviderClient with an valid token and catalog for access to openstack services,
+// or an error if the authentication couldn't be resolved.
+func (c *AccessConfig) Auth() (*gophercloud.ProviderClient, error) {
+	// Note: Some of the fetched env vars here are a bit odd but needs to be here because of compatibility
+	c.Username = common.ChooseString(c.Username, os.Getenv("SDK_USERNAME"), os.Getenv("OS_USERNAME"))
+	c.UserId = common.ChooseString(c.UserId, os.Getenv("OS_USER_ID"))
+	c.Password = common.ChooseString(c.Password, os.Getenv("SDK_PASSWORD"), os.Getenv("OS_PASSWORD"))
+	c.ApiKey = common.ChooseString(c.ApiKey, os.Getenv("SDK_API_KEY"))
+	c.Project = common.ChooseString(c.Project, os.Getenv("SDK_PROJECT"), os.Getenv("OS_TENANT_NAME"), os.Getenv("OS_PROJECT_NAME"))
+	c.Provider = common.ChooseString(c.Provider, os.Getenv("SDK_PROVIDER"), os.Getenv("OS_AUTH_URL"))
+	c.RawRegion = common.ChooseString(c.RawRegion, os.Getenv("SDK_REGION"), os.Getenv("OS_REGION_NAME"))
+	c.TenantId = common.ChooseString(c.TenantId, c.ProjectId, os.Getenv("OS_TENANT_ID"), os.Getenv("OS_PROJECT_ID"))
+	c.Domain = common.ChooseString(c.Domain, os.Getenv("OS_DOMAIN_NAME"))
+	c.DomainId = common.ChooseString(c.DomainId, os.Getenv("OS_PROJECT_DOMAIN_ID"), os.Getenv("OS_USER_DOMAIN_ID"))
```

Comment:
Gophercloud will validate them based on the identity version found in the auth endpoint.